### PR TITLE
Set of redwax modules (modular/federated PKI Tooling)

### PR DESCRIPTION
**Pull request by dirkx**
Set of apache modules that make up most of a PKI framework; along with things like SCEP (automated rollout to things like cisco routers, iPhones, etc), timestamping and so on.

Note that this is our first attempt at a NixOS module set - any and all comments welcome; e.g. should we do it 1) like this - fairly independent; 2) create a library with just the sha/name/desc per module file or 3) create the common elements in the mother module (mod_ca) and minimise the leaf modules by referring somehow to the mod_ca definitions ?

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [X] macOS
   - [X] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: We don't typically share variables like this across different packages. It might seem tedious for this PR, but it's just a few lines to copy/paste. There are some more advanced techniques you could use to parameterize this so you would only need to set it once, but for a first PR let's keep it simple and just type it in a few times, once per package.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: For packages you should use `pname` in combination with `version` so that `name` is properly generated (as `name = pname + '-' + name`). So in this case you want
```
pname = "mod_ca";
version = "0.2.1";
```
and then `name` can be omitted.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: Typically we put this at the end of the file. Not a strict requirement or anything... but everyone else is doing it :sunglasses:

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: No period at the end of `description` please.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: `version` should only be above with `pname` as mentioned earlier, not here.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: As mentioned above it seems best to set this to `url = "https://redwax.eu/dist/rs/${pname}-${version}.tar.gz";` for this PR.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: `gnused` and `coreutils` won't be required, see below.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: As mentioned earlier `gnused` and `coreutils` aren't needed because they are already provided for you, just like `gcc`, which you didn't have to explicitly list.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: You don't need this as far as I can tell.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: This seems to be a good place to mention that indentation in packages should be a consistent 2 spaces per level of indentation.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: Maybe we could put this in `installFlags` like so:
```
  installFlags = [
    "INCLUDEDIR=$(out)/include"
    "LIBEXECDIR=$(out)/modules"
  ];
```

Note that by convention all `mod_*.so` files should go under `${out}/modules`.

File: pkgs/servers/http/apache-modules/mod_csr/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_csr/default.nix]: Where does this file come from? Why is it needed? What does it do? Should you submit a PR upstream to have this fixed in the source?

File: maintainers/maintainer-list.nix
Comment (aanderse) [inline maintainers/maintainer-list.nix]:
```suggestion
  dirkx = {
```
nitpick: missed a space.

**Review by aanderse — CHANGES_REQUESTED** [review]
Welcome and thanks for deciding to contribute to nixpkgs! We're glad you came. I haven't had a chance to test your code yet, but I have left some suggestions I hope you find helpful. Since this is your first PR I thought that my suggestions might be a little overwhelming so I decided to write some code you can reference when going through my suggestions if you get stuck at any point:

https://github.com/NixOS/nixpkgs/compare/master...aanderse:httpd-redwax?expand=1

Do you have a working `configuration.nix` snippet you used to test this code with that you could share?

Don't hesitate to ask any questions at all.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: Will fix :) copied it from the wrong example I guess :) : )

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: So we've done it in this particular way - as it then cleanly overlays with the Apache-httpd (these are modules for apache) and Apache its default path finds these there. So including it is pretty much no configuration file/change/etc - can even be a cmd line flag from systemd.

Moving to modules means that you can no longer rely on that path - and need to be extra careful when sandboxing.

Or is the pattern in https://github.com/NixOS/nixpkgs/blob/465566948393cf533e3617704d1c4ccc34cf3753/pkgs/desktops/gnome-3/core/gnome-user-share/default.nix - line 46 the correct one -- and should any users of this module not use the default configs - but tweak all the paths explicitly for all their modules ?

And should I change this to modules as you suggest (and also do a pull request for apache ?)

File: pkgs/servers/http/apache-modules/mod_csr/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_csr/default.nix]: That PR is working its way through OpenSSL - basically this is a set of forgotten setters/getters. But introducing them introduced another discussion about 1) if the original API is right. 2) the realisation that a lot more were forgotten and 3) concerns that introducing the setters for all of them will perhaps mean an API breaking change about how X509_ALGO* memory is handled - so this will take more than just a few months to propagate through.   And when it does - we have about 9 releases in the 1.x series that have this forgotten call.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: If I take a look at some standard config shipped with `Debian they use `modules/`:
```
cat /etc/apache2/mods-enabled/php7.3.load
# Conflicts: php5
# Depends: mpm_prefork
LoadModule php7_module /usr/lib/apache2/modules/libphp7.3.so
```

I'm pretty sure I've seen `libexecdir/` or some such too... maybe on `RHEL`?

Also `apacheHttpd` via `services.httpd` on NixOS won't automatically find any of your apache modules which aren't part of the `apacheHttpd` package itself. You need to explicitly tell the `httpd` service about them via the `services.httpd.extraModules` option like so:
```
services.httpd.extraModules = [
  { name = "mod_ca"; path = "${pkgs.apacheHttpdPackages}/modules/mod_ca.so"; }
];
```

Notice that I have to specify the whole path. So you can use either `libexecdir` or `modules`, but most packages use `modules` in nixpkgs. No hard requirement.

What would your PR against apache look like? Are you talking about the package or the service?

PS. Actually typing this out makes me realize how tedious the whole `.extraModules path` thing is... PR incoming are some point to make that easier and less error prone.

File: pkgs/servers/http/apache-modules/mod_csr/default.nix
Comment (aanderse) [reply pkgs/servers/http/apache-modules/mod_csr/default.nix]: Sounds like a valid reason to include it here I guess. Who wrote that file? Can we find it online somewhere? Is any other distro using it? Every time we can fetch a file instead of keeping it in tree that is a win.

And documentation is key. We'll want all the facts documented, link to an upstream discussion/PR, notes as to when the file can be removed, etc... If you disappear for any reason someone else needs to track this.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: Well - it gets messier even - as you also have the include part - and the need for APXS to build the libraries in a way that apache can dlopen load them (i.e. get flags like -fPIC not from CFLAGS but from the CFLAGS and other flags used at the time of the apache compile). 

So APXS, as it comes with apache, has its own hardcode include and libexec locations (it predates pkgconfig)  -- and the includes are needed across package boundaries (in this case - we need the apache include files and the mod_ca files. 

Se we use a bit of a trick here to bypass APXS by a modified Makefile.am. But do not try to second guess APXS (by using normal gcc/clang compiles with identical flags). So as thin as possible.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: Ok - so rather than inherit it from the master / mother module -- we copy it into each of the other 12  sub/add-on modules (and do the same for the other values).  And then maintain all these values manually ?

Comment (aanderse) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: Well I counted 7 other packages... but yes, essentially. I try to keep things as straight forward as possible for first time contributors. There are definitely some nice ways to abstract this out that you'll find around nixpkgs, but unless you really expect `https://redwax.eu/dist/rs/` to change often I don't see it being a problem. And even in that case a tool like `sed` goes a long way.

Feel free to copy/paste from the [code I provided you with.](https://github.com/NixOS/nixpkgs/compare/master...aanderse:httpd-redwax?expand=1)

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: Is this something specific to the apache modules you're packaging, or you mean in general? I haven't ran into any issues with any of the apache modules in nixpkgs that I have used. :thinking:

File: maintainers/maintainer-list.nix
Comment (dirkx) [reply maintainers/maintainer-list.nix]: fixed.

File: pkgs/servers/http/apache-modules/mod_csr/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_csr/default.nix]: I've put the URL of the pull request in; along with the likely version by which this file can be removed and some explanation as comments.

If that is unacceptable - the other option is to wait for version 0.2.2 of RedWax - which will have moved these files into the distribution (until August this was not needed - but as of November  - quite a few distributions have the broken openssl versions as their base).

Comment (aanderse) [reply pkgs/servers/http/apache-modules/mod_csr/default.nix]: > quite a few distributions have the broken openssl versions as their base

Do you mean the switch from `openssl` `1.0.x` to `1.1.x`? We have also switched to `1.1.x` by default, but continue to ship `1.0.x` as well. Just trying to clarify what you mean by "broken".

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: You are right - gnused/coreutils are not needed. Removed

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: So we do something a bit unusual and that is we have a main apache module (mod_ca) and then modules that need both the apache dl/apxs -and- mod-ca. So that means that you are not just runtime pulling dl's from any path - but also includes during build. So this is a bit like why you somethings see a libperl or libpython next to a mod_perl stub.

Regardless - we cannot install them in the APXS provided location for modules -- as this one is private to httpd (/nix/store/m15l5xmr2s02sji9yh91nip6b6hrr8r4-apache-httpd-2.4.41/modules/mod_include.so).

But no problem to change this to modules - and then do the same 'sed' substution as for exampled used in [other places](https://github.com/NixOS/nixpkgs/blob/465566948393cf533e3617704d1c4ccc34cf3753/pkgs/desktops/gnome-3/core/gnome-user-share/default.nix#L45) to adapt it at the using side.

Comment (aanderse) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: I have never used `gnome-user-share` but after quickly looking through the derivation and upstream documentation I'm left with the impression that `gnome-user-share` possibly spawns an `apacheHttpd` executable and provides configuration to instruct `apacheHttpd` where apache modules are. This is very separate from how the `services.httpd` nixos module works.

Maybe @jtojnar or @worldofpeace might have some more insight into this, as the resident `gnome` experts, and all around very bright :bulb: people.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: Ok done that. I've left the other (static) values in as inherit from the master. 

Is that ok - or shall I copy those in too ?

Mother:
```
    meta = with stdenv.lib; {
    description = "RedWax CA service module";

    homepage = "https://redwax.eu";
    license = licenses.asl20;
    platforms = platforms.unix;
    maintainers = with maintainers; [dirkx minfrin mvdb];
  };
```
and for the subs:
```
  meta = with stdenv.lib; {
    description = "RedWax CA service module to handle Certificate Signing Requests";

    inherit(mod_ca.meta) license platforms maintainers homepage;
  };

```

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: fixed.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: fixed everywhere.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: fixed everywhere.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: So we do something a bit unusual and that is we have a main apache module (mod_ca) and then modules that need both the apache dl/apxs -and- mod-ca. So that means that you are not just runtime pulling dl's from any path - but also includes during build. So this is a bit like why you somethings see a libperl or libpython next to a mod_perl stub.

Regardless - we cannot install them in the APXS provided location for modules -- as this one is private to httpd (/nix/store/m15l5xmr2s02sji9yh91nip6b6hrr8r4-apache-httpd-2.4.41/modules/mod_include.so).

But no problem to change this to modules - and then do the same 'SED'

Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: fixed all indentations too.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: I've moved it to modules - and tested that I can tweak both cyrus/imapd and wordpress sufficently to be happy. So no issues. Will be fine & can re-visit once we have done the other modules too.

File: pkgs/servers/http/apache-modules/mod_csr/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_csr/default.nix]: I've put some comments in the default.nix file -- but basically after -1.0.2t getters were introduced and part of the openssl struct's became private. 

Most of this quite esoteric - except for the algo bits around a Certificate Signing Request (which you often need to set as the CSR may request defaults ones not acceptale to the policy of the CA) -- so a fair bit of PKI code cannot be compiled against openssl post 1.0.2t (such as  openssl-1.1.0l -to openssl-1.1.0c).

I'll put the pull request for the small fix in openssl in -- though I guess that it may linger & a bigger setter replacement will be needed (and that would also repair the open-ssh patches for x509).

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: you are quite right - minimized now.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (jtojnar) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: I think there is a value in mirroring the FHS structure inside `$out` so that people coming from other distros can quickly understand the output tree. We have been already doing such transitions (https://github.com/NixOS/nixpkgs/pull/43992, https://github.com/NixOS/nixpkgs/pull/30428). Of course it is harder when no such standard location exists. On Arch [`apache`](https://www.archlinux.org/packages/extra/x86_64/apache/) uses `lib/httpd/modules`, whereas [`lighttpd`](https://www.archlinux.org/packages/extra/x86_64/lighttpd/) uses `lib/lighttpd`. It definitely should not be in `$out/modules`, though.

Indeed `gnome-user-share` runs its own `httpd` instance but I would not say it is that different from NixOS module, it is just another instance of extensibility conflicting with immutable store, a well understood issue. The only difference is that in the module, we are creating a configuration from scratch so we can just put in whatever module paths we want, whereas in `gnome-user-share`, we are re-using the upstream config so we need to patch the paths there to combat the assumption that all modules are in a single directory.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (jtojnar) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]:
```suggestion
           "INCLUDEDIR=${placeholder ''out''}/include"
```

is preferred these days, since the value will be directly substituted by Nix. `$(out)` relies on `make` substituting it and it might be confusing with regards to escaping.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (jtojnar) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: Build-time dependencies like `pkg-config` should go to `nativeBuildInputs`.

**Comment by dirkx** [comment]
Ok - so  just changes this to $out/modules -- but happy to change it back !

As to `to such standard location exists' -- it may be easiest go back to our first cut - and follow what 'apxs -q' tells you -- as that tends to follow the idiosyncrasies of the various (linux) platforms.

Dw.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: done.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: And on Debian it is `lib/apache2/modules`. I suppose Debian wasn't really a great example for me to base this on, given the extremes Debian goes to modify `httpd` from upstream. I would guess `lib/httpd/modules` is probably the most standard path and we should go with that.

@jtojnar maybe I should make a PR to utilize `passthru` and add something like `moduleName` and `modulePath` so the `services.httpd.extraModules` can accept a package and always know what to do, similar to what the `odbc` drivers do. This would be a nice addition to the standardization you propose, all in one PR. What do you think?

Comment (jtojnar) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: Sounds good to me.

**Comment by dirkx** [comment]
Jan - not sure if you noticed it - but there is a slight hidden tweak behind that make which is not default Apache-- the https://source.redwax.eu/projects/RS/repos/mod_ca/browse/Makefile.am has been tweaked to both allow the standard linux tweaking; the normal apache synced approach (*bsd, darwin, macports, etc) and a prefix so that you can have the same path within the NIX store that is unique from apache - but they are still the same structure, etc.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]: Gave each module a hardcoded version of the same values - so no longer inheriting.

**Comment by dirkx** [comment]
All review issues should be address I hope.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [reply pkgs/servers/http/apache-modules/mod_ca/default.nix]:
```suggestion
    "--with-apxs=${apacheHttpd.dev}/bin/apxs"
```
@dirkx this line should only have 4 spaces as it is only 2 levels of indentation in.

File: pkgs/servers/http/apache-modules/mod_pkcs12/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_pkcs12/default.nix]: Consistent spacing is always nice:
```suggestion
  buildInputs = [ mod_ca apr aprutil ];
```

File: pkgs/servers/http/apache-modules/mod_scep/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_scep/default.nix]: @jtojnar I don't see this often... are you OK with this?

Comment (jtojnar) [reply pkgs/servers/http/apache-modules/mod_scep/default.nix]: Only a mild dislike but, as long as it is communicated in `mod_ca` that the attributes are shared with reverse dependencies, it is the maintainer’s pain.

File: pkgs/servers/http/apache-modules/mod_pkcs12/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_pkcs12/default.nix]: fixed (wondering if there is a better indent.pro).

File: pkgs/servers/http/apache-modules/mod_scep/default.nix
Comment (dirkx) [reply pkgs/servers/http/apache-modules/mod_scep/default.nix]: Ok - documented.

File: pkgs/servers/http/apache-modules/mod_pkcs12/default.nix
Comment (jtojnar) [reply pkgs/servers/http/apache-modules/mod_pkcs12/default.nix]: You can try nixpkgs-fmt.

**Comment by dirkx** [comment]
Thanks. Works like a treat.

**Comment by dirkx** [comment]
All issues from review addresed.

File: pkgs/top-level/all-packages.nix
Comment (worldofpeace) [inline pkgs/top-level/all-packages.nix]: whitespace issue here

File: pkgs/top-level/all-packages.nix
Comment (worldofpeace) [inline pkgs/top-level/all-packages.nix]: whitespace on the `=`

File: pkgs/top-level/all-packages.nix
Comment (dirkx) [reply pkgs/top-level/all-packages.nix]: fixed. Did not fix the non redwax pacakges.

File: pkgs/top-level/all-packages.nix
Comment (dirkx) [reply pkgs/top-level/all-packages.nix]: fixed. Did not fix the non redwax pacakges.

**Comment by dirkx** [comment]
all fixed. ready for re-review.

File: pkgs/servers/http/apache-modules/mod_ca/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_ca/default.nix]: I believe @jtojnar wanted you to replace `$(out)` with `${placeholder ''out''}` here as well.

**Review by aanderse — CHANGES_REQUESTED** [review]
Sorry for the delay @dirkx. Things are looking good now. One tiny request, see below. After everything has been approved can you please rewrite the git history such that you have a commit which simply adds yourself to the `maintainers-list.nix` file, and then either a single commit for each package or just a single commit for all the packages. Take a quick look at the contributing notes.

Can you please provide your NixOS configuration you have been testing this with?

Thanks!

**Comment by dirkx** [comment]
Ok will do. And will check if one of my test configs is sufficiently clean/unencumbered to already share.

Dw

**Comment by jonringer** [comment]
i think the wrong branch was pushed

**Comment by aanderse** [comment]
@dirkx can you please take a look at your branch when you get a chance?

**Comment by dirkx** [comment]
Closing this pull request - will move & squash the commits onto a new branch - and resubmit a pull request (See #76417).

**Pull request closed without merging**
